### PR TITLE
#2714 download

### PR DESCRIPTION
**Pull request by iamleeg**
Download endpoint. Streams the results to avoid bloating memory; supports CSV, JSON, and TSV. I didn't support the free-text search from the existing data service because we don't expose it in the curator portal, and it is pretty expensive for us to run "at scale".

### CODE REVIEW

File: data-serving/reusable-data-service/reusable_data_service/controller/case_controller.py [hunk 1]
```diff
@@ -109,20 +100,63 @@ def batch_upsert(self, body: dict):
...
+        separator_method = f"{format}_separator"
+        footer_method = f"{format}_footer"
+
+        if case_ids is not None:
+            case_iterator = self.store.identified_case_iterator(case_ids)
+            count = len(case_ids)
+        else:
+            predicate = CaseController.parse_filter(filter)
+            if predicate is None:
+                raise ValidationError("cannot understand query")
```
Comment (jim-sheldon) [inline data-serving/reusable-data-service/reusable_data_service/controller/case_controller.py:130]: Add the predicate and filter to this exception message?

File: data-serving/reusable-data-service/reusable_data_service/main.py [hunk 2]
```diff
@@ -23,21 +31,70 @@ def list_cases():
...
+        }
+        mime_type = content_types.get(req.get("format"), "text/plain")
+        generator = case_controller.download(
+            req.get("format"), req.get("query"), req.get("caseIds")
+        )
+        return app.response_class(generator(), mimetype=mime_type)
+    except PreconditionUnsatisfiedError as e:
+        return jsonify(""), 400
+    except UnsupportedTypeError as e:
+        return jsonify(""), 415
```
Comment (jim-sheldon) [inline data-serving/reusable-data-service/reusable_data_service/main.py:97]: Add the name of the error and/or some context to the returned JSON?

File: data-serving/reusable-data-service/tests/test_case_controller.py [hunk 3]
```diff
@@ -36,180 +43,276 @@ def batch_upsert(self, cases: List[Case]):
...
 def test_list_cases_should_list_the_cases(case_controller):
     with open("./tests/data/case.minimal.json", "r") as minimal_file:
         case = Case.from_json(minimal_file.read())
         case_controller.store.put_case("foo", case)
         case_controller.store.put_case("bar", case)
-    (response, status) = case_controller.list_cases()
-    assert status == 200
-    assert len(response.json["cases"]) == 2
+    cases = case_controller.list_cases()
+    len(cases.cases) == 2
```
Comment (jim-sheldon) [inline data-serving/reusable-data-service/tests/test_case_controller.py:89]: assert?

**Review by jim-sheldon — APPROVED** [review]

File: data-serving/reusable-data-service/reusable_data_service/controller/case_controller.py (hunk 1 above)
Comment (iamleeg) [reply data-serving/reusable-data-service/reusable_data_service/controller/case_controller.py:131]: I didn't do it as the request already contains the query that wasn't understood (and the predicate is `None`), but can do.

**Pull request merged**